s3/src/main/scala/akka/stream/alpakka/s3/impl/S3Stream.scala

```diff
@@ -279,6 +272,29 @@ import scala.util.{Failure, Success, Try}
       .mapMaterializedValue(_.flatMap(identity)(ExecutionContexts.parasitic))
   }
 
+  private def doGetByRanges(
+                             s3Location: S3Location,
+                             versionId: Option[String],
+                             s3Headers: S3Headers,
+                             contentLength: Long,
+                             rangeSize: Long,
+                             parallelism: Int
+                           ): Source[ByteString, Any] = {
+    val byteRanges = computeByteRanges(contentLength, rangeSize)
+    if (byteRanges.size <= 1) {
+      getObject(s3Location, None, versionId, s3Headers)
+    } else {
+      Source(byteRanges)
+        .zipWithIndex
+        .flatMapMerge(parallelism, brToIdx => {
+          val (br, idx) = brToIdx
+          val endMarker = Source.single(ByteString("$END$"))
+          getObject(s3Location, Some(br), versionId, s3Headers).concat(endMarker).map(_ -> idx)
+        })
+        .statefulMapConcat(RangeMapConcat)
```

Comment:
Composing each range-source with a buffer to allow parallel fetching, and then concatenating the resulting streams to get the resulting bytes out in the right order seems like it would achieve the same but much simpler.

Am I missing something clever that this does?

Reply:
If I understood correctly, you are thinking about `conflate` or something similar to buffer range sources.
Something like:

```scala
getObject(s3Location, Some(br), versionId, s3Headers).conflate(_ ++ _).concat(endMarker).map(_ -> idx)
```

As `flatMapMerge` may emit in any order, I still need the range `idx` to order (possibly buffered) bytes.
So output item of `flatMapMerge` will look like `(ByteString, Long)` and can be in any order.

How can I order them back, without `statefulMapConcat` ? Range2 could emit before range 1 and range2 could be complete before range1.

Reply:
I was thinking something like

```
Source(byteRanges)
  .mapAsync(parallelism)(range => 
    getObjectByRanges(...).buffer(size, Backpressure)
  ).flatMapConcat(identity)
```

Reply:
But ofc that may not be good enough with buffer sized in chunks instead of bytes, we don't have a buffer with weighted size calculation though, maybe batchWeighted could do, not sure.

Reply:
Hmm can't make it work:

Tried with:

```scala
Source(byteRanges)
  .mapAsync(parallelism)(br => Future.successful(
    getObject(s3Location, Some(br), versionId, s3Headers).batchWeighted(rangeSize, _.size, identity)(_ ++ _)
  ))
  .flatMapConcat(identity)
```

and

```scala
Source(byteRanges)
  .mapAsync(parallelism)(br => Future.successful(
    Source.fromMaterializer { case (mat, _) =>
      getObject(s3Location, Some(br), versionId, s3Headers)
        .preMaterialize()(mat)
        ._2
        .batchWeighted(rangeSize, _.size, identity)(_ ++ _)
      }
    ))
    .flatMapConcat(identity)
```

But in both situations, ranges are fetched one by one and download perf looks like `getObject`

Reply:
Ah, ofc, they aren't materialized so they can start consume bytes until flatMapConcat:ed, didn't think of that. Pre-materialization creates a running source but the downstream is not materialized until it is used, so you would need to put the batching before preMaterialize.

Reply:
For the record I created an upstream issue with an idea that could make this kind of thing easier: https://github.com/akka/akka/issues/31958 (continue with the current solution here though)